File: server/src/controllers/userController.js
import User from "../models/user.js"
import { asyncHandler } from "../utils/asyncHandler.js"
import bcrypt from "bcrypt"
import { sendResponse } from "../utils/apiResonse.js";
import { statusType } from "../utils/statusType.js";
import { validatePhone } from "../helper/common.js";
import { createToken } from "../helper/common.js";

const registerUser = asyncHandler(async (req, res) => {
    const { name, phone, pin, role } = req.body;

    if (!name || !phone || !pin || !role) {
        return sendResponse(
            res,
            false,
            null,
            "Fields cannot be empty",
            statusType.BAD_REQUEST
        );
    }

    // Check if user already exists
    let user = await User.findOne({ phone });

    if (user) {
        return sendResponse(
            res,
            false,
            null,
            "User Already Exists, Please Login",
            statusType.BAD_REQUEST
        );
    }

    // Validate phone format
    if (!validatePhone(phone)) {
        return sendResponse(
            res,
            false,
            null,
            "Phone number must be a valid phone number",
            statusType.BAD_REQUEST
        );
    }

    // Create new user
    user = new User({ name, phone, pin, role });

    // Hash PIN
    const salt = await bcrypt.genSalt(10);
    user.pin = await bcrypt.hash(pin, salt);

    // Save to DB
    await user.save();

    // Generate token
    const token = createToken(user._id);

    // Prepare response data (exclude PIN)
    const userData = user.toObject();
    delete userData.pin;

    return sendResponse(
        res,
        true,
        { ...userData, token },
        "User Created Successfully",
        statusType.CREATED
    );
});


const loginUser = asyncHandler(async (req, res) => {
  const { phone, pin } = req.body;

  // Check for empty input
  if (!phone || !pin) {
    return sendResponse(res, false, null, "Phone and PIN are required", statusType.BAD_REQUEST);
  }

  // Validate phone number format
  if (!validatePhone(phone)) {
    return sendResponse(res, false, null, "Please enter a valid phone number", statusType.BAD_REQUEST);
  }

  // Find user by phone
  const user = await User.findOne({ phone });

  if (!user) {
    return sendResponse(res, false, null, "User doesn't exist, please create your account", statusType.BAD_REQUEST);
  }

  // Compare PIN
  const isValid = await bcrypt.compare(pin, user.pin);
  if (!isValid) {
    return sendResponse(res, false, null, "Phone or PIN is incorrect", statusType.BAD_REQUEST);
  }

  // Create token
  const token = createToken(user._id);

  // Exclude hashed pin
  const userData = user.toObject();
  delete userData.pin;

  // Send response
  return sendResponse(res, true, { ...userData, token }, "Login Successful", statusType.SUCCESS);
});


export { registerUser, loginUser };

File: server_FastAPI/app/controllers/user_controller.py
from fastapi import APIRouter, Depends, UploadFile, File, HTTPException
from motor.motor_asyncio import AsyncIOMotorDatabase
from app.schemas.user import UserCreate, UserResponse, UserLogin
from app.utils.api_response import ApiResponse
from app.utils.security import create_access_token, create_refresh_token
from app.helper.common import validate_phone
from app.utils.cloudinary import upload_to_cloudinary
from app.db import db
from datetime import datetime
import bcrypt

router = APIRouter()

async def get_user_by_phone(phone: str) -> dict:
    return await db["users"].find_one({"phone": phone})

async def create_user(user_data: dict) -> dict:
    result = await db["users"].insert_one(user_data)
    new_user = await db["users"].find_one({"_id": result.inserted_id})
    return new_user

@router.post("/register", response_model=ApiResponse[UserResponse])
async def register_user(
    user: UserCreate = Depends(),
    image: UploadFile = File(None)
):
    # Phone validation
    if not validate_phone(user.phone):
        return ApiResponse.error("Invalid phone number", 400)
    
    # Check existing user
    if await get_user_by_phone(user.phone):
        return ApiResponse.error("User already exists", 400)
    
    # Hash PIN
    hashed_pin = bcrypt.hashpw(user.pin.encode(), bcrypt.gensalt())
    
    # Upload image if exists
    image_url = ""
    if image:
        image_url = await upload_to_cloudinary(await image.read())
    
    # Create user
    user_data = {
        **user.dict(exclude={"pin"}),
        "pin": hashed_pin.decode(),
        "image": image_url,
        "created_at": datetime.utcnow(),
        "updated_at": datetime.utcnow()
    }
    
    new_user = await create_user(user_data)
    
    # Generate tokens
    access_token = create_access_token(str(new_user["_id"]), new_user["role"])
    refresh_token = create_refresh_token(
        str(new_user["_id"]),
        new_user["role"],
        new_user.get("token_version", 0)
    )
    
    # Update user with refresh token
    await db["users"].update_one(
        {"_id": new_user["_id"]},
        {"$set": {"refresh_token": refresh_token}}
    )
    
    # Prepare response
    response_data = {
        **UserResponse(**new_user).dict(),
        "access_token": access_token,
        "refresh_token": refresh_token
    }
    
    return ApiResponse.success(
        data=response_data,
        message="User registered successfully"
    )

@router.post("/login", response_model=ApiResponse[UserResponse])
async def login_user(credentials: UserLogin):
    # Phone validation
    if not validate_phone(credentials.phone):
        return ApiResponse.error("Invalid phone number", 400)
    
    # Find user
    user = await get_user_by_phone(credentials.phone)
    if not user:
        return ApiResponse.error("User not found", 404)
    
    # Verify PIN
    if not bcrypt.checkpw(credentials.pin.encode(), user["pin"].encode()):
        return ApiResponse.error("Invalid credentials", 401)
    
    # Generate tokens
    access_token = create_access_token(str(user["_id"]), user["role"])
    refresh_token = create_refresh_token(
        str(user["_id"]),
        user["role"],
        user.get("token_version", 0)
    )
    
    # Update user with refresh token
    await db["users"].update_one(
        {"_id": user["_id"]},
        {"$set": {"refresh_token": refresh_token}}
    )
    
    # Prepare response
    response_data = {
        **UserResponse(**user).dict(),
        "access_token": access_token,
        "refresh_token": refresh_token
    }
    
    return ApiResponse.success(
        data=response_data,
        message="Login successful"
    )